### PR TITLE
MINOR: restructure Windows to favor immutable implementation

### PR DESCRIPTION
**Pull request by vvcephei**
In #5510, I added an unsafe (but probably fine in practice) equals/hashCode to the Windows hierarchy.

I believe that this approach is better since:
* It includes no non-deprecated mutable state in equals/hashCode
* It avoids the subclass-return-type builder issues
* It restricts the public interface to only the things that need to be known about all Windows. That is, all Windows need to offer a `gracePeriodMs()`. They do *not* all need to offer a builder method to set it. See `UnlimitedWindows`, for which "grace period" is nonsense. It does not need to allow setting it, and it can return `0`, which is always correct. In contrast, consider `UnlimitedWindows#until(long)`, which has to throw a runtime exception because the interface unnecessarily requires this method to be implemented.

In general, this PR sets a direction in which, once we drop the deprecated members from `Windows`, all the members will be abstract, and `Windows` itself will basically be an interface, which offers a much cleaner and more consistent implementation target.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by vvcephei** [comment]
@guozhangwang @mjsax,

While considering KIP-358, I got an (IMO) much better idea for fixing the problem I fixed in #5510 .

What do you think about it?

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java]: We can't remove the retention time implementation in the superclass, but we don't actually have to use it, either. Since `maintainDurationMs` in the superclass is private, we can simply override the setter and getter and instead implement it as an immutable field.

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java]: Unfortunately, `segments` is public, so we can't bypass it.

File: streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java]: I can't think of a good reason for time- and join-windows to share this default, so they may as well just declare it separately. This is similar reasoning to why `SessionWindows` declares its default separately.

Note that `UnlimitedWindows`, which is also a subclass of `Windows`, has a different "default" retention period (`Long.MAX_VALUE`). Further evidence that it's actually not right to declare it in the abstract superclass.

File: streams/src/main/java/org/apache/kafka/streams/kstream/UnlimitedWindows.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/UnlimitedWindows.java]: this warning probably means we need an integration test

File: streams/src/main/java/org/apache/kafka/streams/kstream/UnlimitedWindows.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/UnlimitedWindows.java:127]: As I said earlier, this change really shines in `UnlimitedWindows` - it's not forced to override methods that don't make sense, just to throw exceptions. Now, the API will simply not allow you to set grace period on UnlimitedWindows, making this a compile-time check instead of a runtime one.

File: streams/src/test/java/org/apache/kafka/streams/kstream/UnlimitedWindowsTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/UnlimitedWindowsTest.java:55]: Rather than throw an exception, it's just not possible to write this code.

**Comment by guozhangwang** [comment]
@vvcephei I saw that `grace()` function is no longer with `Windows`, but each of its implemented class will just have their only `grace()`, is this intentional?

File: streams/src/test/java/org/apache/kafka/streams/kstream/WindowsTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/WindowsTest.java:63]: Similar to `UnlimitedWindows`, `TestWindows` just doesn't need to provide an API to set the grace period.

**Comment by vvcephei** [comment]
@guozhangwang yes, it is intentional. Actually, not all of the implementations *want* to implement grace (see `UnlimitedWindows`), and now they are not required to.

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java]: nit: `super.segments` to make clear it's inherited?

Or maybe: add a package private constructor to `Windows` to accept `segment` parameter?

File: streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java]: > This is similar reasoning to why SessionWindows declares its default separately.

Something I don't like -- I think we should use the same default for all windows (that need one) and thus share the default value as a single constant -- I would rather prefer to create an internal class that just define this single const and use for all window implementations.

I by the argument to not add it to `Windows` as `UnlimitedWindows` does not have a maintain duration. Thus, adding a new internal class for adding the constant that time/join/session window can just use without inheriting anything.

File: streams/src/main/java/org/apache/kafka/streams/kstream/UnlimitedWindows.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/UnlimitedWindows.java:127]: Why return zero ? What about `Long.MAX_VALUE`?

**Review by mjsax — COMMENTED** [review]
I think this changes makes sense. Some minor comments

@guozhangwang Can you have a look? Note, that if we accept this change, we would update KIP-328 accordingly to include difference the public API changes. Not sure if we need to re-cast out votes, but even this would not be hard.

File: streams/src/main/java/org/apache/kafka/streams/kstream/UnlimitedWindows.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/UnlimitedWindows.java:127]: It's immaterial, since `UnlimitedWindows` have no end.

I chose `0` since that was the static result of the prior code: `retention - size = MAX_LONG - MAX_LONG = 0`

**Comment by guozhangwang** [comment]
> @guozhangwang yes, it is intentional. Actually, not all of the implementations want to implement grace (see UnlimitedWindows), and now they are not required to.

Cool. The change lgtm. @mjsax feel free to merge, and @vvcephei needs to update the KIP once it is merged.

**Comment by vvcephei** [comment]
Thanks @guozhangwang . I have a couple of comments to address from @mjsax .

**Comment by vvcephei** [comment]
I've addressed @mjsax 's feedback, as well as @big-andy-coates 's from https://github.com/apache/kafka/pull/5510#discussion_r210384604 (by adding a general equality check)

File: streams/src/test/java/org/apache/kafka/streams/EqualityCheck.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/EqualityCheck.java:19]: This allows us to verify all the properties we wish to hold w.r.t. equals/hashCode.

Most of them are required by either the Java spec or by common sense:
https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#equals(java.lang.Object)
https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#hashCode()

The one "extra" requirement is that inequal objects should have different hashcodes. The spec specifically says this is allowed, but it will cause poor map/set performance, since they will be guaranteed to collide and wind up creating linked lists.
Since we're only validating our own implementations, it's ok to insist on this property.

File: streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:652]: I've added this semi-randomly because I noticed we had 0 unlimited windows in integration tests at all. If you can think of some place you'd rather test it, I'm happy to redo it.

**Review by mjsax — APPROVED** [review]
LGTM.

**Comment by mjsax** [comment]
@vvcephei Can you update the KIP and follow up on the VOTE thread with a brief summary of the changes. If there are no concerns raised on the mailing list we can go ahead and merge this PR -- don't want to merge it before its put out to the mailing list. Thx.

**Comment by vvcephei** [comment]
@mjsax , I did so 19h ago. I'm not sure if that's long enough.

**Comment by tedyu** [comment]
lgtm

File: streams/src/test/java/org/apache/kafka/streams/kstream/WindowsTest.java
Comment (kamalcph) [inline streams/src/test/java/org/apache/kafka/streams/kstream/WindowsTest.java]: nit: unused variable

File: streams/src/test/java/org/apache/kafka/streams/EqualityCheck.java
Comment (kamalcph) [inline streams/src/test/java/org/apache/kafka/streams/EqualityCheck.java]: nit: `[%d]!=[%d].`

File: streams/src/test/java/org/apache/kafka/streams/EqualityCheck.java
Comment (kamalcph) [inline streams/src/test/java/org/apache/kafka/streams/EqualityCheck.java]: nit: `[%d]!=[%d]`

File: streams/src/test/java/org/apache/kafka/streams/EqualityCheck.java
Comment (kamalcph) [inline streams/src/test/java/org/apache/kafka/streams/EqualityCheck.java]: nit: o2[%s] was **equal** to o1[%s]

File: streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java
Comment (kamalcph) [inline streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java]: nit: Same producerConfig is used in all the 4 places. Could you please reuse the same ?

File: streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java
Comment (kamalcph) [inline streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java]: can you document/comment why grace is set to null and segments are not passed here ?

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java
Comment (kamalcph) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java]: can you document/comment why grace is set to null and segments are not passed here ?

**Review by kamalcph — COMMENTED** [review]
LGTM, left some minor comments.

File: streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java]: This is a static method, so grace, segments, and maintainDurationMs are not defined in this scope (since they are instance fields).

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java]: @mjsax @tedyu , thanks for the reviews.

@kamalcph , Good catches! I've addressed them.

**Comment by vvcephei** [comment]
@mjsax @tedyu , thanks for the reviews.

@kamalcph , Good catches! I've addressed them.

File: streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java
Comment (kamalcph) [reply streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindows.java]: Thanks for the explanation. I missed that this method is `static`.

**Comment by mjsax** [comment]
Will merge if Jenkins is green.

**Comment by vvcephei** [comment]
Test failure was: kafka.admin.ListConsumerGroupTest.testListConsumerGroups

Retest this, please.

**Pull request merged**
